File: src/personalities/ambiverts/recent.ts
import { invoke } from "@tauri-apps/api/core";
import { logger } from "../../utility/logger";
import type { VideoItem } from "../../types/core";

/*
dusty::api::recent::get_recent_episodes,
dusty::api::recent::add_recent_episode,
dusty::api::recent::reset_recent_episodes_table,
*/

const CMD_GET_RECENT_VIEW_VIDEOS = 'get_recent_episodes';
const CMD_PUT_EPISODE_IN_RECENT = 'add_recent_episode';
const CMD_RESET_RECENT_EPISODES_TABLE = 'reset_recent_episodes_table';

export async function getRecentViewedEpisodesIPC(): Promise<VideoItem[]> {
    try {
        let result = await invoke<VideoItem[]>(CMD_GET_RECENT_VIEW_VIDEOS);
        return result;
    } catch (error) {
        logger.error(`getRecentViewedEpisodesIPC error: ${error}`);
        return [];
    }
}

export async function putEpisodeInRecentIPC(video: VideoItem): Promise<boolean> {
    try {
        let result = await invoke<boolean>(CMD_PUT_EPISODE_IN_RECENT, { video: video });
        return result;
    } catch (error) {
        logger.error(`putEpisodeInRecentIPC error: ${error}`);
        return false;
    }
}

export async function resetRecentEpisodesTableIPC(): Promise<boolean> {
    try {
        let result = await invoke<boolean>(CMD_RESET_RECENT_EPISODES_TABLE);
        return result;
    } catch (error) {
        logger.error(`resetRecentEpisodesTableIPC error: ${error}`);
        return false;
    }
}

File: src-tauri/src/dusty/api/recent/recent.rs
use crate::dusty::models::state::AppState;
use crate::dusty::db::recent::{
    VideoItem, add_recent_episode_in_db, get_recent_episodes_from_db, reset_recent_episodes_table_in_db,
};

use crate::dusty::logger::logger;

#[tauri::command]
pub async fn add_recent_episode(
    state: tauri::State<'_, AppState>,
    video: VideoItem,
) -> Result<(), String> {
    state
        .db_worker
        .run(move |conn| {
            add_recent_episode_in_db(conn, video).map_err(|e| {
                logger::error!("ADD_RECENT_EPISODE_FAILED", e.log_details());
                e.to_user_message()
            })
        })
        .await
        .map_err(|e| e)?
}

#[tauri::command]
pub async fn get_recent_episodes(
    state: tauri::State<'_, AppState>,
) -> Result<Vec<VideoItem>, String> {
    state
        .db_worker
        .run(|conn| {
            get_recent_episodes_from_db(conn).map_err(|e| {
                logger::error!("GET_RECENT_EPISODES_FAILED", e.log_details());
                e.to_user_message()
            })
        })
        .await
        .map_err(|e| e)?
}

#[tauri::command]
pub async fn reset_recent_episodes_table(
    state: tauri::State<'_, AppState>,
) -> Result<(), String> {
    state
        .db_worker
        .run(|conn| {
            reset_recent_episodes_table_in_db(conn).map_err(|e| {
                logger::error!("RESET_RECENT_EPISODES_TABLE_FAILED", e.log_details());
                e.to_user_message()
            })
        })
        .await
        .map_err(|e| e)?
}
